assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java

```diff
@@ -3215,38 +3252,173 @@ protected String build_instructor_grade_submission_context(VelocityPortlet portl
                     checkForUsersInMultipleGroups(a, s.getSubmitterIds(), state, rb.getString("group.user.multiple.warning"));
                 }
 		
-		// for the navigation purpose
-		List<SubmitterSubmission> userSubmissions = state.getAttribute(USER_SUBMISSIONS) != null ? (List<SubmitterSubmission>) state.getAttribute(USER_SUBMISSIONS):null;
+		// SAK-29314
+		// Since USER_SUBMISSIONS is restricted to the page size, it is not very useful here
+		// Therefore, we will prefer to use STATE_PAGEING_TOTAL_ITEMS. However, sometimes this contains
+		// Assignment objects instead of SubmitterSubmission objects, so we have to fall back to USER_SUBMISSIONS
+		// if this occurs, although in practise this seems to never happen
+		List<SubmitterSubmission> userSubmissions = Collections.EMPTY_LIST;
+		Object totalItems = state.getAttribute(STATE_PAGEING_TOTAL_ITEMS);
+		if (totalItems != null)
+		{
+			List itemList = (List) totalItems;
+			if (!itemList.isEmpty())
+			{
+				if (itemList.get(0) instanceof SubmitterSubmission)
+				{
+					userSubmissions = (List<SubmitterSubmission>) itemList;
+				}
+			}
+		}
+
+		if (userSubmissions.isEmpty())
+		{
+			userSubmissions = state.getAttribute(USER_SUBMISSIONS) != null ? (List<SubmitterSubmission>) state.getAttribute(USER_SUBMISSIONS) : null;
+		}
+
+		// SAK-29314
+		resetNavOptions();
 		if (userSubmissions != null)
 		{
-			for (int i = 0; i < userSubmissions.size(); i++)
+			for( SubmitterSubmission submission : userSubmissions )
 			{
-				if (((SubmitterSubmission) userSubmissions.get(i)).getSubmission().getId().equals(submissionId))
+				if( submission.getSubmission().getId().equals( submissionId ) )
 				{
+					// Determine next/previous
+					int currentIndex = userSubmissions.indexOf( submission );
 					boolean goPT = false;
 					boolean goNT = false;
-					if ((i - 1) >= 0)
+					if( currentIndex > 0 )
 					{
 						goPT = true;
 					}
-					if ((i + 1) < userSubmissions.size())
+					if( currentIndex < userSubmissions.size() - 1 )
 					{
 						goNT = true;
 					}
-					context.put("goPTButton", Boolean.valueOf(goPT));
-					context.put("goNTButton", Boolean.valueOf(goNT));
 					
-					if (i>0)
+					// Determine next ungraded
+					for( int i = currentIndex + 1; i < userSubmissions.size(); i++ )
+					{
+						if( isUngraded( userSubmissions.get( i ).getSubmission(), true ) )
```

Comment:
method naming is a little unclear here, at first glance it looks like nothing is being done after the isUngraded check, but actually the isUngraded method is doing the work. The "is" part of the name makes me think this is a simple state check. Perhaps something more like "processIfUngraded"?